driver/ppm_api_version.h

```diff
@@ -0,0 +1,68 @@
+#ifndef PPM_API_VERSION_H
+#define PPM_API_VERSION_H
+
+/*
+ * API version component macros
+ *
+ * The version is a single uint64_t, structured as follows:
+ * bit 63: unused (so the version number is always positive)
+ * bits 44-62: major version
+ * bits 24-43: minor version
+ * bits 0-23: patch version
+ */
+
+/* extract components from an API version number */
+#define PPM_API_VERSION_MAJOR(ver) ((((ver) >> 44)) & (((1 << 19) - 1)))
+#define PPM_API_VERSION_MINOR(ver) (((ver) >> 24) & (((1 << 20) - 1)))
+#define PPM_API_VERSION_PATCH(ver) (((ver) & ((1 << 24) - 1)))
+
+/* build an API version number from components */
+#define PPM_API_VERSION(major, minor, patch) \
+	(((major) & (((1ULL << 19) - 1) << 44)) | \
+	((minor) & (((1ULL << 20) - 1) << 24)) | \
+	((major) & (((1ULL << 24) - 1))))
+
+#define PPM_API_CURRENT_VERSION PPM_API_VERSION( \
+	PPM_API_CURRENT_VERSION_MAJOR, \
+	PPM_API_CURRENT_VERSION_MINOR, \
+	PPM_API_CURRENT_VERSION_PATCH)
+
+#define __PPM_STRINGIFY1(x) #x
+#define __PPM_STRINGIFY(x) __PPM_STRINGIFY1(x)
+
+#define PPM_API_CURRENT_VERSION_STRING \
+	__PPM_STRINGIFY(PPM_API_CURRENT_VERSION_MAJOR) "." \
+	__PPM_STRINGIFY(PPM_API_CURRENT_VERSION_MINOR) "." \
+	__PPM_STRINGIFY(PPM_API_CURRENT_VERSION_PATCH)
+
+#endif
+
+enum version_check_result {
+	COMPATIBLE,
+	WARN_COMPATIBLE,
+	INCOMPATIBLE,
+};
+
+static inline enum version_check_result check_api_compatibility(unsigned long probe_api_version)
+{
+	if(PPM_API_VERSION_MAJOR(probe_api_version) != PPM_API_CURRENT_VERSION_MAJOR)
+	{
+		// major numbers disagree
+		return INCOMPATIBLE;
+	}
+
+	if(PPM_API_VERSION_MINOR(probe_api_version) < PPM_API_CURRENT_VERSION_MAJOR)
+	{
+		// probe's minor version is < ours
+		return INCOMPATIBLE;
+	}
+	if(PPM_API_VERSION_PATCH(probe_api_version) < PPM_API_CURRENT_VERSION_PATCH)
+	{
+		// probe's patch level is < ours
+		// this is not a fatal error but indicates that the probe has
+		// known issues
+		return WARN_COMPATIBLE;
+	}
```

Comment:
I think there's an issue with the patch version checking here. 
Assuming probe = `1.1.0` and ours = `1.0.5` the last _if_ will return `WARN_COMPATIBLE` though it shouldn't.

A more precise check should be:
```
	if(PPM_API_VERSION_MINOR(probe_api_version) == PPM_API_CURRENT_VERSION_MINOR
          && PPM_API_VERSION_PATCH(probe_api_version) < PPM_API_CURRENT_VERSION_PATCH)
```

Moreover, since client implementation MAY rely on buggy behaviors too, I would not introduce any warning. So I would just treat it as an incompatible version, as semver does.

Thus, my suggestion is:

```suggestion
        if(PPM_API_VERSION_MINOR(probe_api_version) == PPM_API_CURRENT_VERSION_MINOR
          && PPM_API_VERSION_PATCH(probe_api_version) < PPM_API_CURRENT_VERSION_PATCH)
	{
		// probe's patch level is < ours
		return INCOMPATIBLE;
	}
```

Reply:
> Assuming probe = 1.1.0 and ours = 1.0.5 the last if will return WARN_COMPATIBLE though it shouldn't.

Right. Thanks again for stepping up where my brain obviously failed :)

> So I would just treat it as an incompatible version, as semver does.

Hmm. There's something in this that makes me feel uncomfortable but I can't quite put my finger on it. Though I'll just probably do what you suggest.